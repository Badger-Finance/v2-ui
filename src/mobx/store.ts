import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
