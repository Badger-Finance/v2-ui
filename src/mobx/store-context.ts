import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
