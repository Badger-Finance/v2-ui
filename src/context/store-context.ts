import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;