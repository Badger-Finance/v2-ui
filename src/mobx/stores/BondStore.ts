import { extendObservable } from 'mobx';
import { RootStore } from 'mobx/RootStore';
import { DEBUG } from '../../config/environment';
import { Beneficiary, BondType, CitadelBond, IBond } from 'pages/CitadelEarlyBonding/bonds.config';
import { CitadelSale__factory, CitadelSaleGuestList__factory, ERC20__factory } from 'contracts';
import { TokenBalance } from 'mobx/model/tokens/token-balance';
import { ETH_DEPLOY } from 'mobx/model/network/eth.network';
import BigNumber from 'bignumber.js';
import { BigNumber as BigNumberEthers, ethers } from 'ethers';
import { NETWORK_IDS } from '../../config/constants';
import { LOCAL_DEPLOY } from '../model/network/local.network';
import { CitadelMerkleClaim } from '@badger-dao/sdk';

interface CitadelBondInfo {
	tokenRatio: number;
	tokenPrice: number;
	purchasedTokens: number;
	purchasedTokensValue: number;
	purchasedBonds: number;
	purchasedBondsValue: BigNumber;
}

function resolveBondAddress(address?: string): string {
	return address ? address : ethers.constants.AddressZero;
}

/**
 * TODO: Clean up general ethers contract call support
 */
export class BondStore {
	public bonds: CitadelBond[];
	private merkleProof?: CitadelMerkleClaim | null;
	private guestLists = new Map<string, boolean>();

	constructor(private store: RootStore) {
		this.bonds = [];
		extendObservable(this, {
			bonds: this.bonds,
		});
	}

	get allBonds(): IBond[] {
		const { network } = this.store.network;

		if (network.id === NETWORK_IDS.LOCAL) {
			return [
				{
					bondType: BondType.Reserve,
					bondAddress: resolveBondAddress(LOCAL_DEPLOY.citadel?.testSale),
				},
				{
					bondType: BondType.Reserve,
					bondAddress: resolveBondAddress(LOCAL_DEPLOY.citadel?.testSale2),
				},
			];
		}

		return [
			{ bondType: BondType.Reserve, bondAddress: resolveBondAddress(ETH_DEPLOY.citadel?.wbtcSale) },
			{ bondType: BondType.Reserve, bondAddress: resolveBondAddress(ETH_DEPLOY.citadel?.cvxSale) },
			{
				bondType: BondType.Liquidity,
				bondAddress: resolveBondAddress(ETH_DEPLOY.citadel?.bveCVXSale),
			},
			{
				bondType: BondType.Reserve,
				bondAddress: resolveBondAddress(ETH_DEPLOY.citadel?.bcrvibBTCSale),
			},
		];
	}

	isWhitelisted(address: string): boolean {
		return this.guestLists.get(ethers.utils.getAddress(address)) ?? true;
	}

	async updateBonds() {
		const { onboard, sdk } = this.store;
		const { ethersWeb3Provider, address } = onboard;
		if (!ethersWeb3Provider || !address) {
			return;
		}

		const loadedBonds: CitadelBond[] = [];
		await Promise.all(
			this.allBonds.map(async (b) => {
				try {
					const contract = CitadelSale__factory.connect(b.bondAddress, ethersWeb3Provider.getSigner());
					const [
						token,
						ended,
						finalized,
						price,
						userPurchased,
						totalPurchased,
						totalSold,
						claimed,
						guestList,
					] = await Promise.all([
						contract.tokenIn(),
						contract.saleEnded(),
						contract.finalized(),
						contract.tokenOutPrice(),
						contract.boughtAmounts(address),
						contract.totalTokenIn(),
						contract.totalTokenOutBought(),
						contract.hasClaimed(address),
						contract.guestlist(),
					]);

					if (guestList !== ethers.constants.AddressZero) {
						const isWhitelisted = await this.checkWhitelist(guestList);
						this.guestLists.set(ethers.utils.getAddress(b.bondAddress), isWhitelisted);
					}

					const bondToken = await sdk.tokens.loadToken(token);
					const bond = {
						address: token,
						bondToken,
						bondAddress: b.bondAddress,
						price,
						start: BigNumberEthers.from('0'),
						finalized,
						ended,
						bondType: b.bondType,
						userPurchased,
						totalPurchased,
						totalSold,
						claimed,
					};
					loadedBonds.push(bond);
				} catch (err) {
					if (DEBUG) {
						console.error(`Failed to load ${b.bondAddress} ${b.bondType} bond!`);
						console.error(err);
					}
				}
			}),
		);
		this.bonds = loadedBonds;
	}

	async bond(bond: CitadelBond, amount: TokenBalance, beneficiary: Beneficiary): Promise<void> {
		const { onboard } = this.store;
		const { ethersWeb3Provider, address } = onboard;
		if (!ethersWeb3Provider || !address) {
			return;
		}
		console.log(`Triggered a bonding event for ${bond.address} (${amount.balanceDisplay()})`);
		const beneficiaryId = 0; // TODO: look up from some set table
		const sale = CitadelSale__factory.connect(bond.bondAddress, ethersWeb3Provider.getSigner());
		if (!sale) {
			if (DEBUG) {
				throw Error('Sale contract not defined for bond');
			}
			return;
		}
		const bondProof: string[] = []; // look up proof from user store account details
		try {
			const bondingToken = ERC20__factory.connect(bond.address, ethersWeb3Provider.getSigner());
			const allowance = await bondingToken.allowance(address, bond.bondAddress);
			// console.log({ allowance: allowance.toString(), amount: amount.tokenBalance.toString() })
			if (allowance.lt(amount.tokenBalance.toString())) {
				await bondingToken.approve(bond.bondAddress, ethers.constants.MaxUint256);
			}
			const tx = await sale.buy(amount.tokenBalance.toString(), beneficiaryId, bondProof);
			await tx.wait();
			await Promise.all([this.updateBonds(), this.store.user.reloadBalances()]);
		} catch (err) {
			if (DEBUG) {
				console.log(err);
			}
		}
	}

	async claim(bond: CitadelBond): Promise<void> {
		const { onboard } = this.store;
		const { ethersWeb3Provider, address } = onboard;
		if (!ethersWeb3Provider || !address) {
			return;
		}
		console.log(`Triggered a claim event for ${bond.address}`);
		const sale = CitadelSale__factory.connect(bond.bondAddress, ethersWeb3Provider.getSigner());
		if (!sale) {
			if (DEBUG) {
				throw Error('Sale contract not defined for bond');
			}
			return;
		}
		try {
			const tx = await sale.claim();
			await tx.wait();
			await Promise.all([this.updateBonds(), this.store.user.reloadBalances()]);
		} catch (err) {
			if (DEBUG) {
				console.log(err);
			}
		}
	}

	getBondInfo(bond: CitadelBond): CitadelBondInfo {
		let tokenRatio = 0;
		if (bond.price) {
			const basePrice = Number(ethers.utils.formatUnits(bond.price, bond.bondToken.decimals));
			tokenRatio = 1 / basePrice;
		}

		const { bondToken } = bond;
		// const bondedTokenPrice = prices.getPrice(bond.bondToken.address);
		const bondTokenPrice = this.store.prices.getPrice(ETH_DEPLOY.tokens.wBTC);
		const tokenPrice = Number(bondTokenPrice.toString()) / tokenRatio;

		// citadel token has 9 decimals - we can dynamically update this potentially
		const purchasedTokens = Number(ethers.utils.formatUnits(bond.userPurchased, 9));
		const purchasedTokensValue = tokenPrice * purchasedTokens;
		const purchasedBonds = Number(ethers.utils.formatUnits(bond.totalPurchased, bondToken.decimals));
		const purchasedBondsValue = bondTokenPrice.multipliedBy(purchasedBonds);

		return {
			tokenPrice,
			tokenRatio,
			purchasedTokens,
			purchasedTokensValue,
			purchasedBonds,
			purchasedBondsValue,
		};
	}

	private async checkWhitelist(guestList: string): Promise<boolean> {
		const { ethersWeb3Provider } = this.store.onboard;
		const merkleProof = await this.getMerkleProof();

		if (!merkleProof || !ethersWeb3Provider) {
			return false;
		}

		const guestListContract = CitadelSaleGuestList__factory.connect(guestList, ethersWeb3Provider.getSigner());
		return guestListContract.authorized(guestList, merkleProof.proof);
	}

	private async getMerkleProof(): Promise<CitadelMerkleClaim | null | undefined> {
		const { address } = this.store.onboard;

		if (!address) {
			return;
		}

		if (this.merkleProof !== undefined) {
			return this.merkleProof;
		}

		try {
			const merkleProof = await this.store.sdk.api.loadCitadelMerkleProof(address);
			// method will return an empty object on 404
			this.merkleProof = Object.keys(merkleProof).length > 0 ? merkleProof : null;
		} catch (err) {
			console.error(err);
		}

		return this.merkleProof;
	}
}

export default BondStore;
