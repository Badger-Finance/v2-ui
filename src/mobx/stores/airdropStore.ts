import { extendObservable, action } from 'mobx';
import Web3 from 'web3';
import BigNumber from 'bignumber.js';
import { RootStore } from '../RootStore';
import { AbiItem } from 'web3-utils';
import { getSendOptions, sendContractMethod } from 'mobx/utils/web3';
import { AirdropMerkleClaim } from 'mobx/model/rewards/airdrop-merkle-claim';
import { fetchData } from '../../utils/fetchData';
import { DEBUG } from '../../config/environment';

export interface AirdropInformation {
	token: string;
	amount: BigNumber;
	proof: AirdropMerkleClaim;
	airdropAbi: AbiItem[];
	airdropContract: string;
}

class AirdropStore {
	private store!: RootStore;

	public airdrops: AirdropInformation[];

	constructor(store: RootStore) {
		this.store = store;
		this.airdrops = [];

		extendObservable(this, {
			airdrops: this.airdrops,
		});
	}

	fetchAirdrops = action(async () => {
		const { provider, connectedAddress } = this.store.wallet;
		const { network } = this.store.network;
		if (!connectedAddress || !network.airdrops) {
			return;
		}

		const web3 = new Web3(provider);
		this.airdrops = [];

		// For each active airdrop in airdrops.ts:
		// Call API to get merkle proof
		// Check if claimed
		// Set airdrop
		await Promise.all(
			network.airdrops.map(async (airdrop) => {
				if (!airdrop.active) {
					return;
				}

				const [proof] = await fetchData<AirdropMerkleClaim>(`${airdrop.endpoint}/${connectedAddress}`);

				if (!proof) {
					if (DEBUG) {
						this.store.uiState.queueNotification('Unable to retrieve airdrop proof!', 'error');
					}

					return;
				}

				const contract = new web3.eth.Contract(airdrop.airdropAbi, airdrop.airdropContract);
				const claimed = await contract.methods.isClaimed(proof.index).call();

				if (!claimed[0]) {
					this.airdrops.push({
						token: airdrop.token,
						amount: new BigNumber(proof.amount),
						airdropContract: airdrop.airdropContract,
						airdropAbi: airdrop.airdropAbi,
						proof: proof,
					});
				}
			}),
		);
	});

	claimAirdrops = action(
		async (airdropContract: string, airdropAbi: AbiItem[], claim: AirdropMerkleClaim): Promise<void> => {
			const { provider, connectedAddress } = this.store.wallet;
			const { queueNotification, gasPrice } = this.store.uiState;
			const { gasPrices, network } = this.store.network;

			if (!connectedAddress || !network.airdrops || !claim.proof) {
				return;
			}

			const web3 = new Web3(provider);
			const airdropTree = new web3.eth.Contract(airdropAbi, airdropContract);
			const method = airdropTree.methods.claim(claim.index, connectedAddress, claim.amount, claim.proof);

			queueNotification(`Sign the transaction to claim your airdrop`, 'info');
			if (!gasPrices || !gasPrices[gasPrice]) {
				queueNotification(
					`Error retrieving gas selection - check the gas selector in the top right corner.`,
					'error',
				);
				return;
			}

			const price = gasPrices[gasPrice];
			const options = await getSendOptions(method, connectedAddress, price);
			await sendContractMethod(this.store, method, options, `Claim submitted.`, `Aidrop claimed.`);
		},
	);
}

export default AirdropStore;
