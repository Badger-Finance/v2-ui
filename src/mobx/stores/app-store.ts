import { extendObservable, action } from 'mobx';
import Web3 from 'web3'
import { OpenSeaPort, Network } from 'opensea-js'
import { OpenSeaAsset, OrderSide } from 'opensea-js/lib/types';
import { TimerSharp } from '@material-ui/icons';
import BatchCall from "web3-batch-call";
import { collections } from '../../config/constants';
import { reduceVaults } from './reducers';
import _ from 'lodash'
import { estimateAndSend } from "./web3"
import BigNumber from 'bignumber.js';
import { PromiEvent } from 'web3-core';
import { Contract } from 'web3-eth-contract';


const MAX_UINT256 = new BigNumber(2)
	.pow(256)
	.minus(1)
	.toFixed(0);

const ERC20 = require("../../config/abis/ERC20.json")

const infuraProvider = new Web3.providers.HttpProvider('https://mainnet.infura.io/v3/77a0f6647eb04f5ca1409bba62ae9128')
const options = {
	web3: new Web3(infuraProvider),
	etherscan: {
		apiKey: "NXSHKK6D53D3R9I17SR49VX8VITQY7UC6P",
		delayTime: 300
	},
}

const batchCall = new BatchCall(options);


class AppStore {
	public title?: string;
	public user?: any;
	public collection?: any;
	public vault?: any;
	public assets?: any;
	public provider?: any;
	public errorMessage?: String;


	constructor() {
		const provider = new Web3.providers.HttpProvider('https://mainnet.infura.io/v3/77a0f6647eb04f5ca1409bba62ae9128')

		extendObservable(this, {
			title: 'Common Yield',
			user: undefined,
			collection: undefined,
			vault: undefined,
			assets: {},
			provider: undefined,
			errorMessage: undefined,
		});
	}

	setTitle = action((title: string | undefined) => {
		this.title = title;
	});

	setProvider = action((provider: any) => {
		this.provider = provider;
		!!this.collection && this.fetchCollection(this.collection.config.id)
		this.fetchBalancesAndAllowances();
	});
	refresh = action(() => {
		this.setProvider(this.provider)
	});


	fetchCollection = action((collection: string, callback?: () => any) => {

		let theCollection = collections.find((c) => c.id == collection)

		let addressMethods = !!theCollection!.addressMethods && !!this.provider &&
			theCollection!.addressMethods!.map((method: string) => {
				return {
					name: method,
					args: [
						this.provider.selectedAddress
					]
				}
			});

		const contracts = [
			{
				namespace: "vaults",
				abi: theCollection!.abi,
				addresses: theCollection!.vaults,
				allReadMethods: true,
				groupByNamespace: true,
				logging: false,

			} as any
		];
		if (!!addressMethods)
			contracts[0].readMethods = addressMethods

		batchCall.execute(contracts)
			.then((result: any) => {
				const vaults = reduceVaults(result)

				let config = (!!this.collection && this.collection.config.id == theCollection!.id) ? this.collection.config : theCollection
				this.collection = { vaults, config }

				this.fetchBalancesAndAllowances(callback)

			})
			.catch((error: any) => console.log(error))

	});

	fetchBalancesAndAllowances = action((callback?: () => any) => {

		let addresses = this.collection.vaults.map((vault: any) => [vault[this.collection.config.config.underlying!], vault[this.collection.config.config.yielding!]]);
		addresses = _.uniq(_.flatten(addresses))

		const allowances = this.collection.vaults.map((vault: any) => {
			return {
				name: "allowance",
				args: [
					!!this.provider ? this.provider.selectedAddress : '0x0000000000000000000000000000000000000000',
					vault.address,
				]
			}
		});

		const balances = [
			{
				namespace: "balances",
				abi: ERC20.abi,
				addresses,
				allReadMethods: true,
				groupByNamespace: true,
				logging: false,
				readMethods: [{
					name: "balanceOf",
					args: [
						!!this.provider ? this.provider.selectedAddress : '0x0000000000000000000000000000000000000000'
					]
				},
				...allowances]
			}
		];

		batchCall.execute(balances)
			.then((result: any) => {
				console.log(result)
				const bals = reduceVaults(result)

				this.assets = _.keyBy(bals, 'address')

				!!callback && callback()

			}).catch((error: any) => this.errorMessage = error.message)


	});

	fetchVault = action((collection: string, id: string) => {
		console.log("fetching")
		if (!this.collection)
			this.fetchCollection(collection, () => {
				this.vault = this.collection.vaults.find((v: any) => v.address == id)
			})
		else
			this.vault = this.collection.vaults.find((v: any) => v.address == id)

	});

	callMethod = action((address: string, methodName: string, inputs: any = [], abi: any, callback: (contract: PromiEvent<Contract>) => void) => {
		const web3 = new Web3(this.provider)
		const contract = new web3.eth.Contract(abi, address)

		const method = contract.methods[methodName](...inputs)

		estimateAndSend(web3, method, this.provider.selectedAddress, (transaction: PromiEvent<Contract>) => {
			callback(transaction)
		})

	});

	increaseAllowance = action(() => {
		const underlying = this.vault[this.collection.config.config.underlying]

		if (!underlying)
			return

		const underlyingAsset = this.assets[underlying]

		const web3 = new Web3(this.provider)
		const underlyingContract = new web3.eth.Contract(ERC20.abi, underlying)
		const method = underlyingContract.methods.approve(this.vault.address, underlyingAsset.totalSupply)

		estimateAndSend(web3, method, this.provider.selectedAddress, (transaction: PromiEvent<Contract>) => {
			transaction
				.on('transactionHash', (hash: string) => {
					this.errorMessage = hash
				}).on('receipt', (reciept: any) => {
					this.errorMessage = "Allowance increased."
				}).catch((error: any) => this.errorMessage = error.message)

		})
	});

	// UI
	addFilter = action((filter: string) => {
		this.collection.config.config.table.push(filter)
	});
	removeFilter = action((filter: string) => {
		this.collection.config.config.table = this.collection.config.config.table.filter((item: string) => item != filter)
	});
	addAction = action((method: string) => {
		this.collection.config.config.actions.push(method)
	});
	removeAction = action((method: string) => {
		this.collection.config.config.actions = this.collection.config.config.actions.filter((item: string) => item != method)
	});




}

export default AppStore;