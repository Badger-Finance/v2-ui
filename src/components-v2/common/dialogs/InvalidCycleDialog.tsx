import {
	Box,
	Dialog,
	DialogContent,
	DialogContentText,
	DialogTitle,
	IconButton,
	Link,
	Tooltip,
	Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import BugReportIcon from '@material-ui/icons/BugReport';
import CloseIcon from '@material-ui/icons/Close';
import copy from 'copy-to-clipboard';
import { StoreContext } from 'mobx/stores/store-context';
import { observer } from 'mobx-react-lite';
import React, { useContext, useState } from 'react';

import { TOOLTIP_LEAVE_TOUCH_DELAY } from '../../../config/constants';
import CopyToClipboardIcon from '../CopyToClipboardIcon';

interface Props {
	open: boolean;
	onClose: () => void;
}

const useStyles = makeStyles((theme) => ({
	title: {
		paddingBottom: 8,
	},
	content: {
		paddingTop: 0,
	},
	logo: {
		width: 250,
		display: 'block',
		margin: '0 auto',
	},
	bugIcon: {
		marginLeft: theme.spacing(0.5),
	},
	closeButton: {
		position: 'absolute',
		top: theme.spacing(1),
		right: theme.spacing(2),
	},
	code: {
		color: '#b7b9bc',
		pageBreakInside: 'avoid',
		fontFamily: 'monospace',
		fontSize: '15px',
		display: 'flex',
		padding: 4,
		wordWrap: 'break-word',
		alignItems: 'center',
	},
	copyToClipboardButton: {
		marginLeft: theme.spacing(0.5),
		color: theme.palette.common.white,
	},
	copyToClipboardIcon: {
		width: '1em',
		height: '1em',
	},
}));

const InvalidCycleDialog = ({ open, onClose }: Props): JSX.Element => {
	const {
		tree,
	} = useContext(StoreContext);
	const [showCopiedText, setShowCopiedText] = useState(false);
	const classes = useStyles();

	const copyToClipboard = () => {
		if (!tree.cycle) return;
		const didCopy = copy(tree.cycle.toString());
		setShowCopiedText(didCopy);
	};

	return (
		<Dialog open={open} onClose={onClose} fullWidth maxWidth="md">
			<DialogTitle className={classes.title}>
				<Box display="flex" alignItems="center">
					Invalid Cycle Detected <BugReportIcon className={classes.bugIcon} />
					<IconButton aria-label="close dialog" className={classes.closeButton} onClick={onClose}>
						<CloseIcon />
					</IconButton>
				</Box>
			</DialogTitle>
			<DialogContent className={classes.content}>
				<img className={classes.logo} src="/assets/icons/badger-error.png" alt="badger error logo" />
				<DialogContentText id="alert-dialog-description" color="textPrimary">
					The rewards could not be claimed because the current cycle is invalid. Please report this in the ‚ùì
					<Link
						target="_blank"
						href="https://discord.com/channels/743271185751474307/767912941106233355"
						rel="noopener"
					>
						support channel.
					</Link>
				</DialogContentText>
				{tree.cycle && (
					<pre className={classes.code}>
						<Typography color="textPrimary">Cycle: {tree.cycle}</Typography>
						<Tooltip title={showCopiedText ? 'Copied!' : 'Copy to clipboard'} placement="right-start" arrow>
							<IconButton
								onClick={copyToClipboard}
								className={classes.copyToClipboardButton}
								aria-label="copy to clipboard"
								onMouseEnter={() => setShowCopiedText(false)}
								onMouseLeave={() => {
									setTimeout(() => {
										setShowCopiedText(false);
									}, TOOLTIP_LEAVE_TOUCH_DELAY);
								}}
							>
								<CopyToClipboardIcon className={classes.copyToClipboardIcon} />
							</IconButton>
						</Tooltip>
					</pre>
				)}
			</DialogContent>
		</Dialog>
	);
};

export default observer(InvalidCycleDialog);
