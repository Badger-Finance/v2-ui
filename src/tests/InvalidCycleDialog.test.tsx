import copy from 'copy-to-clipboard';
import { StoreProvider } from 'mobx/stores/store-context';
import React from 'react';

import InvalidCycleDialog from '../components-v2/common/dialogs/InvalidCycleDialog';
import store from '../mobx/stores/RootStore';
import { customRender, fireEvent, screen } from './Utils';

jest.mock('copy-to-clipboard');

describe('InvalidCycleDialog', () => {
  beforeEach(() => {
    store.wallet.address = '0x1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a1a';
    store.tree.cycle = 123;
  });

  it('displays invalid cycle dialog', () => {
    const { baseElement } = customRender(
      <StoreProvider value={store}>
        <InvalidCycleDialog open={true} onClose={jest.fn()} />
      </StoreProvider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('can copy cycle to clipboard', () => {
    customRender(
      <StoreProvider value={store}>
        <InvalidCycleDialog open={true} onClose={jest.fn()} />
      </StoreProvider>,
    );
    fireEvent.click(screen.getByRole('button', { name: 'copy to clipboard' }));
    expect(copy).toHaveBeenCalledWith('123');
  });
});
