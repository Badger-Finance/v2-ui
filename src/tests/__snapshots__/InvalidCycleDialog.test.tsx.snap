// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvalidCycleDialog displays invalid cycle dialog 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root-9"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root-24"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container-12 MuiDialog-scrollPaper-10"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root-26 MuiDialog-paper-13 MuiDialog-paperScrollPaper-14 MuiDialog-paperWidthMd-19 MuiDialog-paperFullWidth-22 MuiPaper-elevation24-53 MuiPaper-rounded-27"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root-54 makeStyles-title-1"
        >
          <h2
            class="MuiTypography-root-55 MuiTypography-h6-65"
          >
            <div
              class="MuiBox-root-85 MuiBox-root-86"
            >
              Invalid Cycle Detected 
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-87 makeStyles-bugIcon-4"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 8h-2.81c-.45-.78-1.07-1.45-1.82-1.96L17 4.41 15.59 3l-2.17 2.17C12.96 5.06 12.49 5 12 5c-.49 0-.96.06-1.41.17L8.41 3 7 4.41l1.62 1.63C7.88 6.55 7.26 7.22 6.81 8H4v2h2.09c-.05.33-.09.66-.09 1v1H4v2h2v1c0 .34.04.67.09 1H4v2h2.81c1.04 1.79 2.97 3 5.19 3s4.15-1.21 5.19-3H20v-2h-2.09c.05-.33.09-.66.09-1v-1h2v-2h-2v-1c0-.34-.04-.67-.09-1H20V8zm-6 8h-4v-2h4v2zm0-4h-4v-2h4v2z"
                />
              </svg>
              <button
                aria-label="close dialog"
                class="MuiButtonBase-root-105 MuiIconButton-root-96 makeStyles-closeButton-5"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label-104"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-87"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root-128"
                />
              </button>
            </div>
          </h2>
        </div>
        <div
          class="MuiDialogContent-root-108 makeStyles-content-2"
        >
          <img
            alt="badger error logo"
            class="makeStyles-logo-3"
            src="/assets/icons/badger-error.png"
          />
          <p
            class="MuiTypography-root-55 MuiDialogContentText-root-110 MuiTypography-body1-57 MuiTypography-colorTextPrimary-80"
            id="alert-dialog-description"
          >
            The rewards could not be claimed because the current cycle is invalid. Please report this in the ‚ùì
            <a
              class="MuiTypography-root-55 MuiLink-root-111 MuiLink-underlineHover-113 MuiTypography-colorPrimary-78"
              href="https://discord.com/channels/743271185751474307/767912941106233355"
              rel="noopener"
              target="_blank"
            >
              support channel.
            </a>
          </p>
          <pre
            class="makeStyles-code-6"
          >
            <p
              class="MuiTypography-root-55 MuiTypography-body1-57 MuiTypography-colorTextPrimary-80"
            >
              Cycle: 
              123
            </p>
            <button
              aria-label="copy to clipboard"
              class="MuiButtonBase-root-105 MuiIconButton-root-96 makeStyles-copyToClipboardButton-7"
              tabindex="0"
              title="Copy to clipboard"
              type="button"
            >
              <span
                class="MuiIconButton-label-104"
              >
                <svg
                  class="makeStyles-copyToClipboardIcon-8"
                  fill="none"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9.59981 15.9991H1.60075C1.38867 16.0064 1.17734 15.9701 0.979913 15.8923C0.78248 15.8145 0.603163 15.6969 0.453112 15.5469C0.30306 15.3968 0.185484 15.2175 0.107694 15.0201C0.0299036 14.8227 -0.00643645 14.6113 0.000933297 14.3993V6.40019C-0.00643645 6.18811 0.0299036 5.97679 0.107694 5.77935C0.185484 5.58192 0.30306 5.4026 0.453112 5.25255C0.603163 5.1025 0.78248 4.98492 0.979913 4.90713C1.17734 4.82934 1.38867 4.793 1.60075 4.80037H4.80037V1.60075C4.793 1.38867 4.82934 1.17734 4.90713 0.979913C4.98492 0.78248 5.1025 0.603163 5.25255 0.453112C5.4026 0.30306 5.58192 0.185484 5.77935 0.107694C5.97679 0.0299036 6.18811 -0.00643645 6.40019 0.000933297H14.3993C14.6113 -0.00643645 14.8227 0.0299036 15.0201 0.107694C15.2175 0.185484 15.3968 0.30306 15.5469 0.453112C15.6969 0.603163 15.8145 0.78248 15.8923 0.979913C15.9701 1.17734 16.0064 1.38867 15.9991 1.60075V9.59981C16.0063 9.81186 15.9699 10.0231 15.8921 10.2205C15.8142 10.4179 15.6967 10.5972 15.5466 10.7472C15.3966 10.8972 15.2173 11.0148 15.02 11.0926C14.8226 11.1705 14.6113 11.2069 14.3993 11.1996H11.1996V14.3993C11.2069 14.6113 11.1705 14.8226 11.0926 15.02C11.0148 15.2173 10.8972 15.3966 10.7472 15.5466C10.5972 15.6967 10.4179 15.8142 10.2205 15.8921C10.0231 15.9699 9.81186 16.0063 9.59981 15.9991ZM1.60075 6.40019V14.3993H9.59981V11.1996H6.40019C6.18814 11.2069 5.97687 11.1705 5.77949 11.0926C5.58211 11.0148 5.40284 10.8972 5.25282 10.7472C5.10279 10.5972 4.98521 10.4179 4.90738 10.2205C4.82954 10.0231 4.79312 9.81186 4.80037 9.59981V6.40019H1.60075ZM6.40019 1.60075V9.59981H14.3993V1.60075H6.40019Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root-128"
              />
            </button>
          </pre>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
